server/src/main/java/org/jboss/as/server/BootstrapImpl.java

```diff
@@ -116,52 +115,14 @@ public BootstrapImpl() {
-        tracker.addService(Services.JBOSS_AS, applicationServerService)
-            .install();
-        final ServiceController<?> rootService = container.getRequiredService(Services.JBOSS_AS);
-        rootService.addListener(new AbstractServiceListener<Object>() {
-            @Override
-            public void transition(final ServiceController<?> controller, final ServiceController.Transition transition) {
-                switch (transition) {
-                    case STARTING_to_UP: {
-                        controller.removeListener(this);
-                        final ServiceController<?> controllerServiceController = controller.getServiceContainer().getRequiredService(Services.JBOSS_SERVER_CONTROLLER);
-                        controllerServiceController.addListener(new AbstractServiceListener<Object>() {
-                            public void transition(final ServiceController<?> controller, final ServiceController.Transition transition) {
-                                switch (transition) {
-                                    case STARTING_to_UP: {
-                                        future.done();
-                                        controller.removeListener(this);
-                                        break;
-                                    }
-                                    case STARTING_to_START_FAILED: {
-                                        future.failed(controller.getStartException());
-                                        controller.removeListener(this);
-                                        break;
-                                    }
-                                    case REMOVING_to_REMOVED: {
-                                        future.failed(ServerLogger.ROOT_LOGGER.serverControllerServiceRemoved());
-                                        controller.removeListener(this);
-                                        break;
-                                    }
-                                }
-                            }
-                        });
-                        break;
-                    }
-                    case STARTING_to_START_FAILED: {
-                        controller.removeListener(this);
-                        future.failed(controller.getStartException());
-                        break;
-                    }
-                    case REMOVING_to_REMOVED: {
-                        controller.removeListener(this);
-                        future.failed(ServerLogger.ROOT_LOGGER.rootServiceRemoved());
-                        break;
-                    }
-                }
-            }
-        });
+        tracker.addService(Services.JBOSS_AS, applicationServerService).install();
+        final FutureServiceContainer future = new FutureServiceContainer(container);
+        try {
+            container.awaitStability();
+            future.done();
```

Comment:
I don't think this provides the same behavior. We're done when Services.JBOSS_SERVER_CONTROLLER is up, not when the ServiceContainer is stable. ServerService.start() (aka Services.JBOSS_SERVER_CONTROLLER ) spawns the boot thread which begins asychronously doing work that eventually installs services. If MSC stabilizes before that thread starts installing, then this will be close to being done when Services.JBOSS_SERVER_CONTROLLER is up. But if it doesn't, then it won't stabilize until the work done by the boot thread reaches a stable point.